crates/tracing-instrumentation/src/lib.rs

```diff
@@ -285,3 +341,173 @@ impl Drop for TracingGuard {
         }
     }
 }
+
+/// invocation_span macro create a span given invocation_id and invocation_target. The created span will show up
+/// mainly in services tracing. It will also show up in runtime traces but in relation to other runtime spans
+/// that are created normally with tracing::span!
+///
+/// level: [`Level`]
+/// prefix: static span name
+/// id: ref to an instance of [`InvocationId`]
+/// target: ref to an instance of [`InvocationTarget`]
+/// tags: is a list of any extra tags that need to be associated with this span for example `tags = (client.ip = "10.20.30.40")`
+/// fields [optional]: is a list of extra custom span builder fields that can be used to override the default ones for example `fields = (with_span_id = 10)`
+#[macro_export]
+macro_rules! invocation_span {
+    (level= $lvl:expr, relation = $relation:expr, prefix= $prefix:expr, id= $id:expr, target= $target:expr, tags=($($($key:ident).+ = $value:expr),*), fields=($($field:ident = $field_value:expr),*)) => {
+        {
+            use opentelemetry::KeyValue;
+
+            let mut attributes = vec![
+                KeyValue::new("rpc.service", $target.service_name().to_string()),
+                KeyValue::new("rpc.method", $target.handler_name().to_string()),
+                KeyValue::new("restate.invocation.target", $target.to_string()),
+            ];
+
+            $(
+                attributes.push(KeyValue::new(stringify!($($key).+), $value));
+            )*
+
+            $crate::invocation_span!(
+                level = $lvl,
+                relation = $relation,
+                prefix = $prefix,
+                id = $id,
+                name = format!("{} {}", $prefix, $target.short()),
+                attributes=attributes,
+                fields=($($field = $field_value),*)
+            )
+        }
+    };
+    (level= $lvl:expr, relation = $relation:expr, prefix= $prefix:expr, id= $id:expr, name= $name:expr, tags=($($($key:ident).+ = $value:expr),*), fields=($($field:ident = $field_value:expr),*)) => {
+        {
+            use opentelemetry::KeyValue;
+
+            let mut attributes = vec![];
+
+            $(
+                attributes.push(KeyValue::new(stringify!($($key).+), $value));
+            )*
```

Comment:
nit: better build the attributes vec within the expression to avoid re-allocations.